# Finding High-Comment-Rates Ways: Exploring  Hacker News Posts

In this project, we'll work with a data set of submissions to popular technology site [Hacker News](https://news.ycombinator.com/)，a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

The data set is [here](https://www.kaggle.com/hacker-news/hacker-news-posts) but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions

# Read in the data and remove the headers.

First, we'll read in the data.

In [1]:
#Read in the data
from csv import reader

open_file=open('hacker_news.csv')
read_file=reader(open_file)
hn=list(read_file)
print(hn[:5])


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
# Remove the headers
headers=hn[0]
hn=hn[1:]
print(headers)
hn[:5]

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

We can see above that the data set contains the title of the posts, the number of comments for each post, and the date the post was created. Let's start by exploring the number of comments for each type of post.

#  Extracting Ask HN and Show HN Posts

Now we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

In [3]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts=[]
show_posts=[]
other_posts=[]

for row in hn:
    title=row[1]
    title=title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


As we can see that there are 1744 **Ask HN** posts and 1162 **Show HN** posts

# Calculating the Average Number of Comments for Ask HN and Show HN Posts

we have separated the "ask posts" and the "show posts" into two list of lists named ask_posts and show_posts. Below are the first five rows in the ask_posts list of lists:

>[
    ['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], 
    ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], 
    ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016', 10:14'], 
  

Below are the first five rows in the show_posts list of lists:

>[
    ['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], 
    ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], 
    ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], 
 

Next, let's determine if ask posts or show posts receive more comments on average

In [4]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comments=0
for row in ask_posts:
    num_comments=row[4]
    num_comments=int(num_comments)
    total_ask_comments=total_ask_comments + num_comments

avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)


14.038417431192661


In [5]:
total_show_comments=0
for row in show_posts:
    num_comments=row[4]
    num_comments=int(num_comments)
    total_show_comments=total_show_comments + num_comments

avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)


10.31669535283993


Above we can see that ask posts receive more comments(14) on average than show posts(10).Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

# Finding the Amount of Ask Posts and Comments by Hour Created

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.

2. Calculate the average number of comments ask posts receive by hour created.

To calculate the amount of ask posts and comments by hour created:

In [6]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt
result_list=[]

for row in ask_posts:
    result_list.append([row[6],int(row[4])])
   
counts_by_hour={}
comments_by_hour={}
date_format="%m/%d/%Y %H:%M"

for row in result_list:
    hour_of_date=row[0]
    comment=row[1]
    #parse the date and create a datetime object
    dt_object=dt.datetime.strptime(hour_of_date, date_format)
    #select just the hour from the datetime object
    dt_hour=dt_object.strftime("%H")
    if dt_hour not in counts_by_hour:
        counts_by_hour[dt_hour]=1
        comments_by_hour[dt_hour]=int(comment)
    else:
        counts_by_hour[dt_hour]+=1
        comments_by_hour[dt_hour]+=int(comment)

comments_by_hour
    


{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

We have created two dictionaries:

* counts_by_hour:contains the number of ask posts created during each hour of the day.
* comments_by_hour:contains the corresponding number of comments ask posts created at each hour received. 

Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [10]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour=[]

for hr in comments_by_hour:
    avg_by_hour.append([hr,comments_by_hour[hr]/counts_by_hour[hr]])
    
avg_by_hour
    

[['05', 10.08695652173913],
 ['10', 13.440677966101696],
 ['16', 16.796296296296298],
 ['21', 16.009174311926607],
 ['02', 23.810344827586206],
 ['00', 8.127272727272727],
 ['11', 11.051724137931034],
 ['03', 7.796296296296297],
 ['12', 9.41095890410959],
 ['23', 7.985294117647059],
 ['13', 14.741176470588234],
 ['18', 13.20183486238532],
 ['19', 10.8],
 ['04', 7.170212765957447],
 ['17', 11.46],
 ['08', 10.25],
 ['15', 38.5948275862069],
 ['06', 9.022727272727273],
 ['20', 21.525],
 ['09', 5.5777777777777775],
 ['14', 13.233644859813085],
 ['22', 6.746478873239437],
 ['01', 11.383333333333333],
 ['07', 7.852941176470588]]

# Sorting and Printing Values from a List of Lists

Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [11]:
swap_avg_by_hour=[]

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
    
print(swap_avg_by_hour)

sorted_swap=sorted(swap_avg_by_hour,reverse=True)

sorted_swap
    

[[10.08695652173913, '05'], [13.440677966101696, '10'], [16.796296296296298, '16'], [16.009174311926607, '21'], [23.810344827586206, '02'], [8.127272727272727, '00'], [11.051724137931034, '11'], [7.796296296296297, '03'], [9.41095890410959, '12'], [7.985294117647059, '23'], [14.741176470588234, '13'], [13.20183486238532, '18'], [10.8, '19'], [7.170212765957447, '04'], [11.46, '17'], [10.25, '08'], [38.5948275862069, '15'], [9.022727272727273, '06'], [21.525, '20'], [5.5777777777777775, '09'], [13.233644859813085, '14'], [6.746478873239437, '22'], [11.383333333333333, '01'], [7.852941176470588, '07']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [ ]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
temple='{hour}: {num:.2f} average comments per post'
for row in sorted_swap[:5]:
    time=dt.datetime.strptime(row[1],'%H').strftime('%H:%M')
    print(temple.format(hour=time,num=row[0]))

The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm  - 4:00 pm ).